Built CRUD operations on Grishma's work.

johnsciutto opened the pull request:
Added instance methods and static methods to use CRUD operations with the user database.

Did my best to build on Grishma's work.

Unit testing for CRUD operations is not present yet (working on it). I just wanted to make sure this (merging into someone else's branch) was working before adding more work.

johnsciutto committed:
Merged User methods created by John with Grishma's

The functions to perform CRUD operations are now available through static and
instance methods of the user model.

johnsciutto committed:
Linting with prettier, no other edits

jessdesr commented on server/userModel/user.js:
```diff
@@ -1,13 +1,196 @@
+userSchema.methods.softDelete = function () {
+  return User.updateOne(
+    { email: this.email },
+    { isActive: false },
```
The user schema does not have isActive defined: you can get around needing to add this to every schema in the database by making a soft deletion plugin!

johnsciutto replied:
Right! I forgot to add that. Thanks for catching it.

Cool, I didn't know that existed. I'll research how to do that.
Should I also add the property manually to the schema in the file?

johnsciutto replied:
Resolved. Pushing now.

jessdesr reviewed (commented):
Great, we should hold off on merging this into Grishma's branch if her PR is going to be up soon @shahgrishma123 ?

johnsciutto committed:
Merged changes from backend-login-signup

johnsciutto committed:
Add softDeletionPlugin to userSchema

johnsciutto committed:
Change name of create user function

johnsciutto committed:
Updated linting in a lot of files. No big changes.

Commit previous to pulling new updates from parent branch.

johnsciutto committed:
After merge with parent branch.

Merge branch 'backend-login-signup' into backend-login-signup-john.

jessdesr commented:
You'll need to resolve these conflicts, this is most likely due to the linting that was performed on your branch

johnsciutto committed:
Completed mongoose configuration and functions.

- Finished and manually tested mongoose methods in online test database.
- Formatting of userData with prettier.

johnsciutto committed:
Change name of encryptedPasswordWithSalt

To verb to reflect the action that the function performs.

johnsciutto committed:
Refactor createNewUserObject

To be a one function trigger to tigger the creation, validation, hashing and
saving of a user in the database.

johnsciutto committed:
Delete unnecesary functions.

Their functionality is already covered by static and instance methods.

johnsciutto committed:
Better database error handling

johnsciutto committed:
Tightening security in generateAuthToken

johnsciutto committed:
Changed registerUser so that it works async.

Also eliminated unnecesary functions and imports.

johnsciutto committed:
Apply prettier to loginController.js

johnsciutto committed:
Refactor loginController to work asynchronously

And with the mongoose static method (findByEmail).

johnsciutto committed:
Include database functions usage

johnsciutto committed:
Add _ notation to private instance methods

johnsciutto committed:
Improve error messages in loginUser

johnsciutto committed:
Small changes and edits

- Changed `User.findByEmail(...)` to make sure it returns an object and not an
  array with a single object.
- Added @private tag to private methods.

johnsciutto committed:
Moved user model to /models/user-model.js

johnsciutto committed:
Moved encryptPasswordWithSalt to /utils/database.js

johnsciutto committed:
Moved authentication utility functions to /utils/

- Created `authentication.js` in `/utils/`.
- Moved `isPasswordValid` and `generateAuthToken`.
- Wrote documentation for both functions.
